kge/util/sampler.py

```diff
@@ -188,6 +210,368 @@ def _filter_and_resample_fast(
+        negative-sampling tensor (i.e., in `self.samples(indexes).view(-1)`).
+
+        """
+        samples = self.samples(indexes)
+        return torch.unique(samples.view(-1), return_inverse=return_inverse)
+
+    def to(self, device) -> "BatchNegativeSample":
+        """Move the negative samples to the specified device."""
+        self.positive_triples = self.positive_triples.to(device)
+        return self
+
+    def score(self, model, indexes=None) -> torch.Tensor:
+        """Score the negative samples for the batch with the provided model.
+
+        If `indexes` is provided, only score the corresponding subset of the batch.
+
+        Returns a chunk_size x num_samples tensor of scores. Here chunk_size corresponds
+        the batch size (if `indexes=None`) or to the number of specified indexes (otherwise).
+
+        Sets the `forward_time` and `prepare_time` attributes.
+        """
+        self.forward_time = 0.0
+        self.prepare_time = 0.0
+
+        # the default implementation here is based on the set of all samples as provided
+        # by self.samples(); get the relavant data
+        slot = self.slot
+        self.prepare_time -= time.time()
+        negative_samples = self.samples(indexes)
+        num_samples = self.num_samples
+        triples = (
+            self.positive_triples[indexes, :] if indexes else self.positive_triples
+        )
+        self.prepare_time += time.time()
+
+        # go ahead and score
+        device = self.positive_triples.device
+        chunk_size = len(negative_samples)
+        scores = None
+        if self._implementation == "triple":
+            # construct triples
+            self.prepare_time -= time.time()
+            triples_to_score = triples.repeat(1, num_samples).view(-1, 3)
+            triples_to_score[:, slot] = negative_samples.contiguous().view(-1)
+            self.prepare_time += time.time()
+
+            # and score them
+            self.forward_time -= time.time()
+            scores = model.score_spo(
+                triples_to_score[:, S],
+                triples_to_score[:, P],
+                triples_to_score[:, O],
+                direction=SLOT_STR[slot],
+            ).view(chunk_size, -1)
+            self.forward_time += time.time()
+        elif self._implementation in ["batch", "all"]:
+            # Score each triples against all unique possible targets, then pick out the
+            # actual scores.
+            self.prepare_time -= time.time()
+            if self._implementation == "all":
+                unique_targets = None  # means all
+                column_indexes = negative_samples.contiguous().view(-1)
+            else:
+                unique_targets, column_indexes = self.unique_samples(
+                    indexes, return_inverse=True
+                )
+            self.prepare_time += time.time()
+
+            # compute all scores for slot
+            self.forward_time -= time.time()
```

Comment:
Should we move the following 15 lines to a little helper function to reduce duplicated code?

Reply:
Sure!